domain/application/service/containerimageresourcestore_test.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+
+	"github.com/juju/testing"
+	jc "github.com/juju/testing/checkers"
+	"go.uber.org/mock/gomock"
+	gc "gopkg.in/check.v1"
+
+	containermetadataresourcetesting "github.com/juju/juju/core/resources/containermetadataresource/testing"
+	resourcetesting "github.com/juju/juju/core/resources/testing"
+	"github.com/juju/juju/domain/application"
+	"github.com/juju/juju/internal/charm/resource"
+	"github.com/juju/juju/internal/docker"
+)
+
+type containerImageResourceStoreSuite struct {
+	testing.IsolationSuite
+	containerImageResourceState *MockContainerImageResourceState
+	imageMetadata               docker.DockerImageDetails
+	file                        io.ReadCloser
+}
+
+var _ = gc.Suite(&containerImageResourceStoreSuite{})
+
+func (s *containerImageResourceStoreSuite) setupMocks(c *gc.C) *gomock.Controller {
+	ctrl := gomock.NewController(c)
+
+	s.containerImageResourceState = NewMockContainerImageResourceState(ctrl)
+
+	return ctrl
+}
+
+func (s *containerImageResourceStoreSuite) SetUpTest(c *gc.C) {
+	s.imageMetadata = docker.DockerImageDetails{
+		RegistryPath: "url@sha256:abc123",
+		ImageRepoDetails: docker.ImageRepoDetails{
+			BasicAuthConfig: docker.BasicAuthConfig{
+				Username: "testuser",
+				Password: "hunter2",
+			},
+		},
+	}
+	data, err := json.Marshal(s.imageMetadata)
+	c.Assert(err, jc.ErrorIsNil)
+	s.file = io.NopCloser(bytes.NewReader(data))
+}
+
+func (s *containerImageResourceStoreSuite) TestContainerImageResourceStorePut(c *gc.C) {
```

Comment:
**todo:** Add error path test (at least those caused by an error from state layer.

Valid for all tests.

Reply:
Added.